File: astra-desktop-mac/astra-desktop/BoincCommands.swift
//
//  BoincCommands.swift
//  astra-desktop
//
//  Created by Soham Kale on 1/10/18.
//  Copyright © 2018 Soham Kale. All rights reserved.
//

import Foundation

enum BoincCommands {
    
    static func lookupAccount(url: String, email: String, password: String) -> String { return self.exec_cmnd(args: ["--lookup_account", url, email, password]) }
    
            
    static func createUser(url: String, email: String, password: String, name: String) -> String { return self.exec_cmnd(args: ["--create_account", url, email, password, name]) }
    
    static func attachProject(url: String, accountKey: String) -> String { return self.exec_cmnd(args: ["--project_attach", url, accountKey]) }
    
    
    static func getCCStatus() -> String { return self.exec_cmnd(args: ["--get_cc_status"]) }
    
    static func getState() -> String { return self.exec_cmnd(args: ["--get_state"]) }
    
    static func getTasks() -> String { return self.exec_cmnd(args: ["--get_tasks"]) }
    
    static func getProjectStatus() -> String { return self.exec_cmnd(args: ["--get_project_status"]) }
    
    static func getDiskUsage() -> String { return self.exec_cmnd(args: ["--get_disk_usage"]) }
    
    
    static func resetProject(url: String) -> String { return self.exec_cmnd(args: ["--project", url, "reset"]) }
    
    static func detachProject(url: String) -> String { return self.exec_cmnd(args: ["--project", url, "detach"]) }
    
    static func suspendProject(url: String) -> String { return self.exec_cmnd(args: ["--project", url, "suspend"]) }
    
    static func resumeProject(url: String) -> String { return self.exec_cmnd(args: ["--project", url, "resume"]) }
    
    static func noMoreWorkProject(url: String) -> String { return self.exec_cmnd(args: ["--project", url, "nomorework"]) }
    
    
    static func alwaysCPU(duration: Int?) -> String { return self.exec_cmnd(args: ["--set_run_mode", "always", (duration == nil) ? "" : String(describing: duration)]) }
    
    static func autoCPU(duration: Int?) -> String { return self.exec_cmnd(args: ["--set_run_mode", "auto", (duration == nil) ? "" : String(describing: duration)]) }
    
    static func neverCPU(duration: Int?) -> String {
        return self.exec_cmnd(args: ["--set_run_mode", "never", (duration == nil) ? "" : String(describing: duration)])
    }
    
    
    static func alwaysGPU(duration: Int?) -> String { return self.exec_cmnd(args: ["--set_gpu_mode", "always", (duration == nil) ? "" : String(describing: duration)]) }
    
    static func autoGPU(duration: Int?) -> String { return self.exec_cmnd(args: ["--set_gpu_mode", "auto", (duration == nil) ? "" : String(describing: duration)]) }
    
    static func neverGPU(duration: Int?) -> String { return self.exec_cmnd(args: ["--set_gpu_mode", "never", (duration == nil) ? "" : String(describing: duration)]) }
    
    
    static func alwaysNetwork(duration: Int?) -> String { return self.exec_cmnd(args: ["--set_gpu_mode", "always", (duration == nil) ? "" : String(describing: duration)]) }
    
    static func autoNetwork(duration: Int?) -> String { return self.exec_cmnd(args: ["--set_gpu_mode", "auto", (duration == nil) ? "" : String(describing: duration)]) }
    
    static func neverNetwork(duration: Int?) -> String { return self.exec_cmnd(args: ["--set_gpu_mode", "never", (duration == nil) ? "" : String(describing: duration)]) }
    
    
    static func quit() -> String { return self.exec_cmnd(args: ["--quit"]) }

}

extension BoincCommands {
    static func exec_cmnd(args: [String]) -> String {
        return BashController.bash(command: "Boinc/boinccmd", arguments: args)
    }
}

File: astra-desktop-cross-platform/src/core/BoincCommands.java
package core;

/**
 * Copyright (c) 2018 Astra International. All rights reserved.
 * Use of this source code is governed by a BSD License (see LICENSE).
 *
 * A command interface for the boincmmd tool with basic utilities to be exposed to user.
 * Created by Soham Kale on 2/20/18
 *
 */
public class BoincCommands {

    static Thread startBoinc() {
        Thread boinc = new Thread(() -> {
            BashClient.bashPersist("./boinc -insecure");
        });
        boinc.start();
        return boinc;
    }


    /* ****************************** GLOBAL PREFERENCES ****************************** */

    static void setRunOnBatteries(boolean opt) {
        XMLController.editTag("run_on_batteries", opt ? "1" : "0");
        globalPrefsOverride();
    }

    static void setRunIfActive(boolean opt) {
        XMLController.editTag("run_if_user_active", opt ? "1" : "0");
        globalPrefsOverride();
    }

    static void setConfigHours(int startHour, int endHour) {
        XMLController.editTag("start_hour", Integer.toString(startHour));
        XMLController.editTag("end_hour", Integer.toString(endHour));
        globalPrefsOverride();
    }

    static void setMaxCPUs(int max) {
        XMLController.editTag("max_cpus", Integer.toString(max));
        globalPrefsOverride();
    }

    static void setDiskMaxPercent(int percent) {
        XMLController.editTag("disk_max_used_pct", Integer.toString(percent));
        globalPrefsOverride();
    }

    static void setRAMMaxUse(int percent) {
        boolean dummyVar = XMLController.getTag("run_if_user_active").equals("1") ?
                XMLController.editTag("ram_max_used_busy_pct", Integer.toString(percent)) :
                XMLController.editTag("ram_max_used_idle_pct", Integer.toString(percent));
        globalPrefsOverride();
    }

    static void setCPUUsage(int percent) {
        XMLController.editTag("cpu_usage_limit", Integer.toString(percent));
        globalPrefsOverride();
    }



    static String getRunIfActive(boolean opt) {
        return XMLController.getTag("run_if_user_active");
    }

    static String[] getConfigHours(int startHour, int endHour) {
        return new String[] {XMLController.getTag("start_hour"), XMLController.getTag("end_hour")};
    }

    static String getMaxCPUs(int max) {
        return XMLController.getTag("max_cpus");
    }

    static String getDiskMaxPercent(int percent) {
        return XMLController.getTag("disk_max_used_pct");
    }

    static String getRAMMaxUse(int percent) {
        return XMLController.getTag("run_if_user_active").equals("1") ?
                XMLController.getTag("ram_max_used_busy_pct") :
                XMLController.getTag("ram_max_used_idle_pct");
    }

    static String getCPUUsage(int percent) {
        return XMLController.getTag("cpu_usage_limit");
    }


    /* *********************************** BOINCCMD *********************************** */

    static String[] lookupAccount(String url, String email, String password) {
        return execCmnd("--lookup_account", url, email, password);
    }

    static String[] createUser(String url, String email, String password, String name)  {
        return execCmnd("--create_account", url, email, password, name);
    }

    static String[] attachProject(String url, String accountKey)  {
        return execCmnd("--project_attach", url, accountKey);
    }


    static String[] getCCStatus()  { return execCmnd("--get_cc_status"); }

    static String[] getState()  { return execCmnd("--get_state"); }

    static String[] getTasks()  { return execCmnd("--get_tasks"); }

    static String[] getProjectStatus()  {
        return execCmnd("--get_project_status");
    }

    static String[] getDiskUsage()  { return execCmnd("--get_disk_usage"); }


    static String[] resetProject(String url)  { return execCmnd("--project", url, "reset"); }

    static String[] detachProject(String url)  { return execCmnd("--project", url, "detach"); }

    static String[] suspendProject(String url)  { return execCmnd("--project", url, "suspend"); }

    static String[] resumeProject(String url)  { return execCmnd("--project", url, "resume"); }

    static String[] noMoreWorkProject(String url)  { return execCmnd("--project", url, "nomorework"); }


    static String[] runBenchmarks()  { return execCmnd("--run_benchmarks"); }


    static String[] setCPU(String setting, int... duration)  {
        return execCmnd("--set_run_mode", setting, (duration == null) ? "" : String.valueOf(duration[0]));
    }

    static String[] setGPU(String setting, int... duration)  {
        return execCmnd("--set_gpu_mode", setting, (duration == null) ? "" : String.valueOf(duration[0]));
    }

    static String[] setNetwork(String setting, int... duration)  {
        return execCmnd("--set_gpu_mode", setting, (duration == null) ? "" : String.valueOf(duration[0]));
    }


    static String[] quit()  { return execCmnd("--quit"); }

    private static String[] globalPrefsOverride()  { return execCmnd("--read_global_prefs_override"); }


    private static String[] execCmnd(String... args) {
        String[] boincArgs = new String[args.length + 1];
        System.arraycopy(args, 0, boincArgs, 1, args.length);
        boincArgs[0] = "./boinccmd";
        return BashClient.bash(boincArgs);
    }

}
